packages/vm/src/evm/evm.ts

```diff
@@ -579,12 +729,168 @@ export default class EVM {
     }
   }
 
+  /**
+   * Executes an EVM message, determining whether it's a call or create
+   * based on the `to` address. It checkpoints the state and reverts changes
+   * if an exception happens during the message execution.
+   */
+
+  async runCall(opts: RunCallOpts): Promise<EVMResult> {
+    let message = opts.message
+    if (!message) {
+      const block = opts.block ?? Block.fromBlockData({}, { common: this._common })
+      this._block = block
+      const txContext: TxContext = {
+        gasPrice: opts.gasPrice ?? BigInt(0),
+        origin: opts.origin ?? opts.caller ?? Address.zero(),
+      }
+      this._tx = txContext
+
+      const caller = opts.caller ?? Address.zero()
+      const value = opts.value ?? BigInt(0)
+      if (opts.skipBalance) {
+        // if skipBalance, add `value` to caller balance to ensure sufficient funds
+        const callerAccount = await this._state.getAccount(caller)
+        callerAccount.balance += value
+        await this._state.putAccount(caller, callerAccount)
+      }
+
+      message = new Message({
+        caller,
+        gasLimit: opts.gasLimit ?? BigInt(0xffffff),
+        to: opts.to,
+        value,
+        data: opts.data,
+        code: opts.code,
+        depth: opts.depth,
+        isCompiled: opts.isCompiled,
+        isStatic: opts.isStatic,
+        salt: opts.salt,
+        selfdestruct: opts.selfdestruct ?? {},
+        delegatecall: opts.delegatecall,
+      })
+    }
+
+    await this._emit('beforeMessage', message)
+
+    if (!message.to && this._common.isActivatedEIP(2929)) {
+      message.code = message.data
+      ;(<any>this._state).addWarmedAddress((await this._generateAddress(message)).buf)
```

Comment:
do we still need this `any` cast on `this._state`?

Reply:
we don't! removed